src/cargo/core/resolver/mod.rs

```diff
@@ -298,109 +244,213 @@ fn activate(mut cx: Box<Context>,
         Method::Everything => None,
     };
 
-    activate_deps(cx, registry, parent, platform, deps.iter(), 0,
-                  &mut |mut cx, registry| {
-        cx.visited.remove(id);
-        finished(cx, registry)
-    })
+    Ok(Some(DepsFrame{
+        parent: parent,
+        remaining_siblings: RcVecIter::new(deps),
+        platform: platform.map(|s| s.to_string()),
+        id: id,
+    }))
 }
 
-/// Activates the dependencies for a package, one by one in turn.
-///
-/// This function will attempt to activate all possible candidates for each
-/// dependency of the package specified by `parent`. The `deps` iterator
-/// provided is an iterator over all dependencies where each element yielded
-/// informs us what the candidates are for the dependency in question.
-///
-/// The `platform` argument is the target platform that the dependencies are
-/// being activated for.
+#[derive(Clone)]
+struct RcVecIter<Elem> {
+    vec: Rc<Vec<Elem>>,
+    next_index: usize,
```

Comment:
It winds up as `self.rest.next().and_then(|i| self.vec.get(i).map(|val| (i, val.clone())))` because of the need to return the index, which may or may not be more readable than the explicit `match`. I've kept the one-liner for now.